internal/contentmanager/cache/cache.go

```diff
@@ -0,0 +1,240 @@
+}
+
+// CloserSyncingSource is a wrapper of the controller-runtime
+// source.SyncingSource that includes methods for:
+//   - Closing the source, stopping it's interaction with the Kubernetes API server and reaction to events
+type CloserSyncingSource interface {
+	source.SyncingSource
+	io.Closer
+}
+
+type sourcerer interface {
+	Source(schema.GroupVersionKind, client.Object, func(context.Context)) (CloserSyncingSource, error)
+}
+
+type cache struct {
+	sources     map[schema.GroupVersionKind]CloserSyncingSource
+	sourcerer   sourcerer
+	owner       client.Object
+	syncTimeout time.Duration
+	mu          sync.Mutex
+}
+
+func NewCache(sourcerer sourcerer, owner client.Object, syncTimeout time.Duration) Cache {
+	return &cache{
+		sources:     make(map[schema.GroupVersionKind]CloserSyncingSource),
+		sourcerer:   sourcerer,
+		owner:       owner,
+		syncTimeout: syncTimeout,
+	}
+}
+
+var _ Cache = (*cache)(nil)
+
+func (c *cache) Watch(ctx context.Context, watcher Watcher, objs ...client.Object) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	gvkSet, err := gvksForObjects(objs...)
+	if err != nil {
+		return fmt.Errorf("getting set of GVKs for managed objects: %w", err)
+	}
+
+	if err := c.removeStaleSources(gvkSet); err != nil {
+		return fmt.Errorf("removing stale sources: %w", err)
+	}
+	return c.startNewSources(ctx, gvkSet, watcher)
+}
+
+func (c *cache) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	errs := []error{}
+	for _, source := range c.sources {
+		if err := source.Close(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+func (c *cache) startNewSources(ctx context.Context, gvks sets.Set[schema.GroupVersionKind], watcher Watcher) error {
+	cacheGvks := c.getCacheGVKs()
+	gvksToCreate := gvks.Difference(cacheGvks)
+
+	type startResult struct {
+		source CloserSyncingSource
+		gvk    schema.GroupVersionKind
+		err    error
+	}
+	startResults := make(chan startResult)
+	wg := sync.WaitGroup{}
+	for _, gvk := range gvksToCreate.UnsortedList() {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			source, err := c.startNewSource(ctx, gvk, watcher)
+			startResults <- startResult{
+				source: source,
+				gvk:    gvk,
+				err:    err,
+			}
+		}()
+	}
+	go func() {
+		wg.Wait()
+		close(startResults)
+	}()
+
+	sourcesErrors := []error{}
+	for result := range startResults {
+		if result.err != nil {
+			sourcesErrors = append(sourcesErrors, result.err)
+			continue
+		}
+
+		err := c.addSource(result.gvk, result.source)
+		if err != nil {
+			// If we made it here then there is a logic error in
+			// calculating the diffs between what is currently being
+			// watched by the cache
+			panic(err)
+		}
+	}
+
+	return errors.Join(sourcesErrors...)
+}
+
+func (c *cache) startNewSource(ctx context.Context, gvk schema.GroupVersionKind, watcher Watcher) (CloserSyncingSource, error) {
+	s, err := c.sourcerer.Source(gvk, c.owner, func(ctx context.Context) {
+		err := c.removeSource(gvk)
```

Comment:
Nit: might be nice to add a comment here (and maybe also in the `sourcerer` interface definition) to explain what this callback function is used for.

Reply:
Added a comment here and in the sourcerer interface in https://github.com/operator-framework/operator-controller/pull/1119/commits/f6f277035a8331898fd304504b30660905797406